blockchain/upgrade.go

```diff
@@ -3363,6 +3453,449 @@ func separateUtxoDatabase(ctx context.Context, db database.DB, utxoDb database.D
+	utxoDbInfoLegacyBucketName := []byte("dbinfo")
+	utxoDbInfoVersionKeyNameV1 := []byte("version")
+	utxoDbInfoCompVerKeyNameV1 := []byte("compver")
+	utxoDbInfoUtxoVerKeyNameV1 := []byte("utxover")
+	utxoDbInfoCreatedKeyNameV1 := []byte("created")
+	utxoSetStateKeyNameV1 := []byte("utxosetstate")
+
+	// ---------------------------------------------------------------------------
+	// The new keys in the UTXO backend start with a serialized prefix consisting
+	// of the key set and version of that key set as follows:
+	//
+	//   <key set><version>
+	//
+	//   Key        Value    Size      Description
+	//   key set    uint8    1 byte    The key set identifier, as defined below
+	//   version    uint8    1 byte    The version of the key set
+	//
+	// The key sets as of this migration are:
+	//   utxoKeySetDbInfo:    1
+	//   utxoKeySetUtxoState: 2
+	//   utxoKeySetUtxoSet:   3
+	//
+	// The versions as of this migration are:
+	//   utxoKeySetDbInfo:    0
+	//   utxoKeySetUtxoState: 1
+	//   utxoKeySetUtxoSet:   3
+	// ---------------------------------------------------------------------------
+	utxoDbInfoVersionKeyNew := []byte("\x01\x00version")
+	utxoDbInfoCompVerKeyNew := []byte("\x01\x00compver")
+	utxoDbInfoUtxoVerKeyNew := []byte("\x01\x00utxover")
+	utxoDbInfoCreatedKeyNew := []byte("\x01\x00created")
+	utxoSetStateKeyNew := []byte("\x02\x01utxosetstate")
+	utxoPrefixUtxoSetV3 := []byte("\x03\x03")
+
+	// moveKey is a helper function that uses an existing UTXO backend transaction
+	// to move a key.
+	moveKey := func(tx UtxoBackendTx, oldKey, newKey []byte) error {
+		serialized, err := tx.Get(oldKey)
+		if err != nil {
+			return err
+		}
+		err = tx.Put(newKey, serialized)
+		if err != nil {
+			return err
+		}
+		return tx.Delete(oldKey)
+	}
+
+	// Move the database info from the legacy bucket to the new keys.
+	bucketId, err := fetchLegacyBucketId(utxoBackend.Get,
+		utxoDbInfoLegacyBucketName)
+	if err != nil {
+		return err
+	}
+	if bucketId != nil {
+		err := utxoBackend.Update(func(tx UtxoBackendTx) error {
+			// Move the database version from the legacy bucket to the new key.
+			oldKey := prefixedKey(bucketId, utxoDbInfoVersionKeyNameV1)
+			err = moveKey(tx, oldKey, utxoDbInfoVersionKeyNew)
+			if err != nil {
+				return fmt.Errorf("error migrating database version: %v", err)
+			}
+
+			// Move the database compression version from the legacy bucket to the new
+			// key.
+			oldKey = prefixedKey(bucketId, utxoDbInfoCompVerKeyNameV1)
+			err = moveKey(tx, oldKey, utxoDbInfoCompVerKeyNew)
+			if err != nil {
+				return fmt.Errorf("error migrating database compression version: %v",
+					err)
+			}
+
+			// Move the database UTXO set version from the legacy bucket to the new
+			// key.
+			oldKey = prefixedKey(bucketId, utxoDbInfoUtxoVerKeyNameV1)
+			err = moveKey(tx, oldKey, utxoDbInfoUtxoVerKeyNew)
+			if err != nil {
+				return fmt.Errorf("error migrating UTXO set version: %v", err)
```

Comment:
Not sure if dcrd meticulously wraps errors, but could use `%w` unless that's not dcrd style.

Reply:
Yeah, errors are generally wrapped and also typically use specific error kinds/codes, but these were a bit looser in the upgrade code.  Updated to use `%w` here.